text/0000-bionic-mixins.md

```diff
@@ -0,0 +1,75 @@
+# Meta
+[meta]: #meta
+- Name: Defining Mixins for io.buildpacks.stacks.bionic
+- Start Date: 2019-12-05
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+The Buildpack API Specification allows stack maintainers to define stack mixins for a given stack.
+The Cloud Native Buildpacks project informally defines stack build and run images with ID `io.buildpacks.stacks.bionic` as containing exactly the set of Ubuntu packages in the `ubuntu:bionic` image on Docker Hub.
+This RFC proposes that we formalize the existing definition of `io.buildpacks.stacks.bionic` and define mixins on that stack to be Ubuntu 18.04 LTS package names (without version or architecture info) from official package mirrors. 
+
+In general, a stack is a contract that should be more like "Ubuntu Bionic" and less like "Platform XYZ's Re-distribution of Ubuntu Binoic."
+Stack images can still be provided by platforms/vendors, but they should use the standard CNB stack contracts where applicable.
+
+# Motivation
+[motivation]: #motivation
+
+This will allow:
+1. Buildpack authors to publish buildpacks that are compatible with more stack images without requiring those authors to explicitly specify vendor-specific stack IDs.
+2. Vendor-specific buildpacks to work together in the same build, as long as those buildpacks support `io.buildpacks.stacks.bionic`.
+3. App developers to create custom stack images that are compatible with any set of buildpacks that support `io.buildpacks.stacks.bionic`.
+
+# Example
+[example]: #example
+
+Say `org.cloudfoundry.stacks.cfbionicstack` is `ubuntu:bionic`, but with `build-essentials` in the build image and `git` in both build and run images.
+
+We would change the image metadata from:
+```
+run stackID: "org.cloudfoundry.stacks.cfbionicstack"
+run mixins: []
+build stackID: "org.cloudfoundry.stacks.cfbionicstack"
+build mixins: []
+```
+To:
+```
+run stackID: "io.buildpacks.stacks.bionic"
+run mixins: ["git"]
+build stackID: "io.buildpacks.stacks.bionic"
+build mixins: ["build:build-essentials", "git"]
```

Comment:
I gave defining a buildpack properly with mixins a shot. It gets complicated. I would bet most buildpack authors don't even know what they depend on that is not in base bionic without some serious work.

I had to build a build/run image that was _just_ bionic to test my buildpack and what it depends on. Then, I had to run as root to install packages and basically play whack-a-mole as I ran into errors while building an example app. What's the incentive for buildpack authors to do this work? How do they know if they are relying on a package's dependency and it gets removed in next version?

Also, all the buildpacks I tried the process on needed `curl` and `ca-certificates`. If you've only ever built buildpacks against CF/HK images, I can imagine you would never even know those packages exist and just assume you don't need to list them. It seems likely that a great majority of the buildpacks would have a lot of the same packages.

Finally, what about packages that the buildpack doesn't rely on itself, but the majority of the target audience for the buildpack would require? This came up when discussing with @hone. He probably has a real concrete example package but say that most of our customers build ruby apps that require `libreadline`. The buildpack itself doesn't _require_ it and it's optional for building ruby itself but the customer's code it is expecting to build most likely would need it. We could list it as a `mixin` for the buildpack but that would essentially tie the buildpack to Heroku's bionic images that include it. Users who build ruby code that doesn't rely on ruby built with `libreadline` can't use the buildpack on another `bionic` image either.